### ISSUE TITLE
Deployment fails

### ISSUE THREAD

**Issue by eviweb** [cmars/juju-charm-gogs#1]
Hi Casey,

I tried to deploy this charm on a local cloud using juju-2.0-rc1 and running `juju deploy cs:~cmars/gogs` results in the following error:

```
ERROR juju.worker.uniter.operation hook "install" failed: exit status 1
unit-gogs-1: 20:52:11 INFO unit.gogs/1.juju-log Reactive main running for hook install
unit-gogs-1: 20:52:11 INFO unit.gogs/1.juju-log Invoking reactive handler: reactive/gogs.py:17:install
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install Traceback (most recent call last):
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/var/lib/juju/agents/unit-gogs-1/charm/hooks/install", line 19, in <module>
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     main()
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/__init__.py", line 78, in main
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     bus.dispatch()
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/bus.py", line 426, in dispatch
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     _invoke(hook_handlers)
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/bus.py", line 417, in _invoke
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     handler.invoke()
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/bus.py", line 291, in invoke
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     self._action(*args)
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/var/lib/juju/agents/unit-gogs-1/charm/reactive/gogs.py", line 23, in install
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     handler.download(INSTALL_URL % version, dest='/opt/gogs.tar.gz')
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/local/lib/python3.4/dist-packages/charmhelpers/fetch/archiveurl.py", line 107, in download
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     response = urlopen(source)
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/lib/python3.4/urllib/request.py", line 161, in urlopen
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     return opener.open(url, data, timeout)
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/lib/python3.4/urllib/request.py", line 469, in open
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     response = meth(req, response)
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/lib/python3.4/urllib/request.py", line 579, in http_response
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     'http', request, response, code, msg, hdrs)
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/lib/python3.4/urllib/request.py", line 507, in error
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     return self._call_chain(*args)
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/lib/python3.4/urllib/request.py", line 441, in _call_chain
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     result = func(*args)
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install   File "/usr/lib/python3.4/urllib/request.py", line 587, in http_error_default
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install     raise HTTPError(req.full_url, code, msg, hdrs, fp)
unit-gogs-1: 20:52:11 INFO unit.gogs/1.install urllib.error.HTTPError: HTTP Error 404: Not Found
```

It seems the default version set in [./config.yaml](https://github.com/cmars/juju-charm-gogs/blob/master/config.yaml) is no more available.

Best regards

Eric

**Comment by eviweb** [comment]
Hi Casey,

the gogs version issue fixed, the deployment of the gogs charm still fails.
But now during the `db-relation-joined` event with the following error: 
```
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined Traceback (most recent call last):
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined   File "/var/lib/juju/agents/unit-gogs-1/charm/hooks/db-relation-joined", line 19, in <module>
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined     main()
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/__init__.py", line 78, in main
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined     bus.dispatch()
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 434, in dispatch
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined     _invoke(other_handlers)
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 417, in _invoke
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined     handler.invoke()
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 291, in invoke
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined     self._action(*args)
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined   File "/var/lib/juju/agents/unit-gogs-1/charm/reactive/gogs.py", line 59, in db_available
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined     'host': db.host(),
unit-gogs-1: 14:13:04 INFO unit.gogs/1.db-relation-joined AttributeError: 'PostgreSQLClient' object has no attribute 'host'
unit-gogs-1: 14:13:04 ERROR juju.worker.uniter.operation hook "db-relation-joined" failed: exit status 1
```

The current revision number of the deployed Postgresql charm is: 105

Eric

**Comment by cmars** [comment]
I may need to update the postgresql relation. Noticed there was a fix in juju-charm-mattermost recently. It's in my queue..

**Comment by eviweb** [comment]
Hi Casey,

I'm currently working on a fix: https://github.com/eviweb/juju-charm-gogs/tree/fix-issue-%231
- **obsolete version number issue:** fixed and I improve a little bit the download url generation to permit the deployment of the latest version     
- **database relation issue:** fixed, grandly inspired by what was done in `juju-charm-mattermost`

Still pending reverse proxying with `haproxy` charm does not work, I've got a HTTP error 500.

Please let me know if something should be improved before I submit a PR.
I'm totally new to juju and charm development, so any advice is welcome.

Best regards

Eric

**Comment by cmars** [comment]
@eviweb I'd recommend using resources instead. See https://github.com/cmars/juju-charm-mattermost, that has a good example of resources and also works with postgresql.

**Comment by eviweb** [comment]
@cmars, the database relation issue was caused by a misuse of the pgsql interface.
eq. `db.host()` instead of `db.master.host` etc...
I will provide a PR that quickly solves this two issues and permits the deployment of this charm again.
I keep your suggestion regarding the use of resources for later.
Eric

**Comment by cmars** [comment]
Fixed by #2. Thanks @eviweb! I've pushed cs:~cmars/gogs-8 with these changes.
